### PR TITLE
GH-2993 Refactored NTriples/NQuads Parser to use buffering

### PR DESCRIPTION
**Pull request by tokovach**
Signed-off-by: Tomas Kovachev <t.kovachev1996@gmail.com>

GitHub issue resolved: #2993 

Briefly describe the changes proposed in this PR:

Refactored NTriples/NQuads Parser to use a buffered reader for processing input to improve performance.

### CODE REVIEW

**Comment by hmottestad** [comment]
The Eclipse ECA check is complaining about the following: 

> Anonymous (tomas.kovachev@****.com) is not covered by the necessary legal agreements in order to proceed.

File: core/rio/nquads/src/main/java/org/eclipse/rdf4j/rio/nquads/NQuadsParser.java [hunk 1]
```diff
@@ -35,143 +26,82 @@
...
-
-	@Override
-	public synchronized void parse(final InputStream inputStream, final String baseURI)
-			throws IOException, RDFParseException, RDFHandlerException {
-		if (inputStream == null) {
-			throw new IllegalArgumentException("Input stream can not be 'null'");
-		}
+	/*--------------*
+	 * Constructors *
+	 *--------------*/
```
Comment (jeenbroekstra) [inline core/rio/nquads/src/main/java/org/eclipse/rdf4j/rio/nquads/NQuadsParser.java:31]: Fwiw we don't really use these kinds of markers anymore as they do more harm than good. Not a big deal, you can leave it in here, but no need to add this in for any future contributions :)

**Review by jeenbroekstra — COMMENTED** [review]
I haven't looked in detail yet but at first glance a good improvement.

@tokovach are you familiar with JMH benchmarks at all? It might be good to use this opportunity to add one or two micro-benchmarks for parser performance and measure some before- and after figures, just so we have numbers to back up that this is more performant.

File: core/rio/nquads/src/main/java/org/eclipse/rdf4j/rio/nquads/NQuadsParser.java (hunk 1 above)
Comment (tokovach) [reply core/rio/nquads/src/main/java/org/eclipse/rdf4j/rio/nquads/NQuadsParser.java:31]: I don't like these markers either, I just noticed they are present for a number of parsers/writers and wanted to be consistent. But I will remove them :)

**Comment by tokovach** [comment]
I am familiar with JMH, I have created an entire module for parser benchmarking, which as discussed with @hmottestad I will submit from a separate PR. Should I link the said PR to this issue?

**Comment by jeenbroekstra** [comment]
> I am familiar with JMH, I have created an entire module for parser benchmarking, which as discussed with @hmottestad I will submit from a separate PR. Should I link the said PR to this issue?

Ah excellent! Yeah, you might as well reuse this ticket for that.

**Review by hmottestad — APPROVED** [review]
I've had a look, but it's quite a lot of code. I've tried to look at everything, and I can't find anything to fault!

LGTM :)

**Review by jeenbroekstra — APPROVED** [review]
LGTM - good improvement, thanks!

**Pull request merged**